File: typescript/src/tools/insights/getSqlInsight.ts
import { InsightGetSqlSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = InsightGetSqlSchema;

type Params = z.infer<typeof schema>;

export const getSqlInsightHandler = async (context: Context, params: Params) => {
	const { query } = params;
	const projectId = await context.getProjectId();

	const result = await context.api.insights({ projectId }).sqlInsight({ query });

	if (!result.success) {
		throw new Error(`Failed to execute SQL insight: ${result.error.message}`);
	}

	if (result.data.length === 0) {
		return {
			content: [
				{
					type: "text",
					text: "Received an empty SQL insight or no data in the stream.",
				},
			],
		};
	}
	return { content: [{ type: "text", text: JSON.stringify(result.data) }] };
};

const definition = getToolDefinition("get-sql-insight");

const tool = (): Tool<typeof schema> => ({
	name: "get-sql-insight",
	description: definition.description,
	schema,
	handler: getSqlInsightHandler,
});

export default tool;

File: python/tools/insights/get_sql_insight.py
from api.client import is_error, is_success
from schema.tool_inputs import InsightGetSqlSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def get_sql_insight_handler(context: Context, params: InsightGetSqlSchema) -> ToolResult:
    project_id = await context.get_project_id()

    insight_result = await context.api.insights(project_id).sql_insight(params.query)

    if is_error(insight_result):
        error_msg = str(insight_result.error)
        error_type = type(insight_result.error).__name__
        if not error_msg or error_msg == "None":
            error_msg = f"Unknown error occurred (type: {error_type})"

        raise Exception(f"Failed to execute SQL insight: {error_msg}")

    assert is_success(insight_result)

    return ToolResult(content=insight_result.data)


def get_sql_insight_tool() -> Tool[InsightGetSqlSchema]:
    definition = get_tool_definition("get-sql-insight")

    return Tool(
        name="get-sql-insight",
        description=definition["description"],
        schema=InsightGetSqlSchema,
        handler=get_sql_insight_handler,
    )
